service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/flight/RetrieveAzureCloudContexts.java

```diff
@@ -0,0 +1,58 @@
+package bio.terra.workspace.service.resource.controlled.cloud.azure.flight;
+
+import static bio.terra.workspace.common.utils.FlightUtils.validateRequiredEntries;
+import static bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys.CONTROLLED_RESOURCES_WITHOUT_REGION;
+import static bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys.WORKSPACE_ID_TO_AZURE_CLOUD_CONTEXT_MAP;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.service.resource.controlled.model.ControlledResource;
+import bio.terra.workspace.service.resource.model.WsmResource;
+import bio.terra.workspace.service.workspace.AzureCloudContextService;
+import com.fasterxml.jackson.core.type.TypeReference;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+/**
+ * Retrieves the azure cloud context of the Terra workspace where the resources with missing regions
+ * are in. The azure cloud context are deserialized and used in {@link
+ * RetrieveAzureResourcesRegionStep}.
+ */
+public class RetrieveAzureCloudContexts implements Step {
+
+  final AzureCloudContextService azureCloudContextService;
+
+  public RetrieveAzureCloudContexts(AzureCloudContextService azureCloudContextService) {
+    this.azureCloudContextService = azureCloudContextService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    validateRequiredEntries(context.getWorkingMap(), CONTROLLED_RESOURCES_WITHOUT_REGION);
+    List<ControlledResource> controlledResources =
+        context.getWorkingMap().get(CONTROLLED_RESOURCES_WITHOUT_REGION, new TypeReference<>() {});
+    Set<UUID> workspaceIds =
```

Comment:
This could be done in a single database transaction in the DAO that would:

1. Make one database query that would return a tuple: (azureCloudContext, controlledResourceWithoutRegion).
2. Stream that to build a map of cloudContext -> region
3. Loop through that to update the region field of the resource

You are basically implementing a hash join over workspaceId in memory that the database has already built.
This would not need to be a flight at all.

The only nasty thing about that approach is calling CRL from inside the DAO. Not a good plan.
However, you could do it in a flight in two steps:
1. Step 1: Query the database as above
2. Step 2: Build the map of context -> region using CRL. Then call DAO with the map and the list. Have it perform the updates in one transaction.

Since this is single-use-code that you have debugged, it is not worth redoing. However, when you find yourself building and probing maps, maybe check if the database can do it for you.